### ISSUE TITLE
Bw-Tree: Heap-Use-After-Free Error

### ISSUE THREAD

**Issue by apavlo** [cmu-db/noisepage#791]
This error showed up after running the TPC-C tests for several hours. I locally modified `ADD_TERRIER_TESTS` in [BuildUtils.cmake](https://github.com/cmu-db/terrier/blob/master/cmake_modules/BuildUtils.cmake#L329) to only select `tpcc_test.cpp`. This makes it so that when you invoke `make unittest` it only runs the TPC-C tests instead of the full suite.

```
$ cmake -DCMAKE_BUILD_TYPE=relwithdebinfo -DTERRIER_BUILD_BENCHMARKS=OFF -DTERRIER_USE_ASAN=yes ..
$ cnt=0 ; while [ true ]; do echo $cnt ; make -j 2 unittest ; test $? -ne 0 && break; cnt=$((cnt + 1)) ; done ; echo $cnt
<SNIP>
Test project /home/pavlo/Documents/Peloton/Github/terrier/build                             
    Start 1: tpcc_test                          
1/1 Test #1: tpcc_test ........................***Failed   66.35 sec                        
                              
0% tests passed, 1 tests failed out of 1                                                            
                                                                                                     
Label Time Summary:                                                                                   
unittest    =  66.35 sec*proc (1 test)                         
                                                                                                       
Total Test time (real) =  66.35 sec
                                                                                                       
The following tests FAILED:                                     
          1 - tpcc_test (Failed)                                                                       
```

Full error output is [attached](https://github.com/cmu-db/terrier/files/4244856/tpcc_test.txt).

```
==11056==ERROR: AddressSanitizer: heap-use-after-free on address 0x60302196d480 at pc 0x56066f4f561c bp 0x7fbadc4f7970 sp 0x7fbadc4f7960
WRITE of size 4 at 0x60302196d480 thread T22
    #0 0x56066f4f561b in std::__atomic_base<int>::fetch_sub(int, std::memory_order) /usr/include/c++/7/bits/atomic_base.h:524
    #1 0x56066f4f561b in third_party::bwtree::BwTree<terrier::storage::index::CompactIntsKey<(unsigned char)16>, terrier::storage::TupleSlot, std::less<terrier::storage::index::CompactIntsKey<(unsigned char)16> >, std::equal_to<terrier::storage::index::CompactIntsKey<(unsigned char)16> >, std::hash<terrier::storage::index::CompactIntsKey<(unsigned char)16> >, std::equal_to<terrier::storage::TupleSlot>, std::hash<terrier::storage::TupleSlot> >::EpochManager::LeaveEpoch(third_party::bwtree::BwTree<terrier::storage::index::CompactIntsKey<(unsigned char)16>, terrier::storage::TupleSlot, std::less<terrier::storage::index::CompactIntsKey<(unsigned char)16> >, std::equal_to<terrier::storage::index::CompactIntsKey<(unsigned char)16> >, std::hash<terrier::storage::index::CompactIntsKey<(unsigned char)16> >, std::equal_to<terrier::storage::TupleSlot>, std::hash<terrier::storage::TupleSlot> >::EpochManager::EpochNode*) /home/pavlo/Documents/Peloton/Github/terrier/third_party/bwtree/bwtree.h:7238
```

**Comment by mbutrovich** [comment]
Dupe of #544?

**Comment by crd477** [comment]
I noticed that @jrolli created https://github.com/wangziqi2013/BwTree/issues/69.  Aside from also being a use-after-free, I don't know if it's related.

**Comment by jrolli** [comment]
I don’t think this is the same race I identified in the upstream issue.  It may be related since that is a GC-related race for the Bwtrees, but at first glance the free and usage points weren’t what I recalled.  I’ll try to take a look and confirm one way or the other since there was a clear way ahead to solve the race I had identified.

**Comment by lmwnshn** [comment]
This will be suppressed (but not fixed) by #1224. The plan is to replace the BwTree eventually.

**Comment by lmwnshn** [comment]
RIP BwTree
